Feature: kubectl create ingress

amimof opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Adds the ability to create ingress resources using `kubectl create ingress` command. This will speed up the development workflow by providing a quick and easy way of deploying *and* accessing apps. 

For example. As a developer I want to deploy my app to an existing staging cluster and be able to access it. In most cases I will be using `kubectl create serviceaccount`, `kubectl create service`, `kubectl create deployment`. But without an ingress I wont be able to access it, and `kubectl create ingress` does not exist currently.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
None

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Adds `create ingress` command to `kubectl` 
```

amimof commented:
/joke

amimof commented:
/assign @seans3

dims commented:
/uncc

amimof commented:
/auto-cc

amimof commented:
/auto-cc

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_ingress.go:
```diff
@@ -0,0 +1,82 @@
+		DisableFlagsInUseLine: true,
+		Aliases:               []string{"ing"},
+		Short:                 i18n.T("Create an ingress with the specified name."),
+		Long:                  templates.LongDesc(i18n.T("Create an ingress with the specified name.")),
```
Long and example are usually defined outside for readability.

amimof replied:
These are now created in two global cars

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_ingress.go:
```diff
@@ -0,0 +1,82 @@
+
+// NewCmdCreateIngress is a macro command to create a new ingress.
+func NewCmdCreateIngress(f cmdutil.Factory, ioStreams genericclioptions.IOStreams) *cobra.Command {
+	options := &IngressOpts{
```
I want to see `NewIngressOptions` method setting this.

amimof replied:
Looked at other commands and the standard seems to be `NewCreateIngressOptions`. Is this correct?

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_ingress.go:
```diff
@@ -0,0 +1,82 @@
+)
+
+// IngressOpts is returned by NewCmdCreateIngress
+type IngressOpts struct {
```
`IngressOptions`

amimof replied:
Renamed

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_ingress.go:
```diff
@@ -0,0 +1,82 @@
+	cmdutil.AddGeneratorFlags(cmd, generateversioned.IngressV1GeneratorName)
+	cmd.Flags().String("host", "", i18n.T("Host name this Ingress should route traffic on"))
+	cmd.Flags().String("service-name", "", i18n.T("Service this Ingress should route traffic to"))
+	cmd.Flags().Int("service-port", 80, "Port number of the Service to route traffic to")
```
Defaults should be set in `NewIngressOptions`.

soltysh replied:
Each flag should be wired with appropriate member of `IngressOptions` struct.

amimof replied:
Fixed this accordingly

amimof replied:
Fixed

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_ingress.go:
```diff
@@ -0,0 +1,82 @@
+			Host:     		cmdutil.GetFlagString(cmd, "host"),
+			ServiceName:  cmdutil.GetFlagString(cmd, "service-name"),
+			ServicePort:  cmdutil.GetFlagInt(cmd, "service-port"),
+			Path:					cmdutil.GetFlagString(cmd, "path"),
```
Formatting is messed up here.

soltysh replied:
You should be using values wired from `IngressOptions` struct.

amimof replied:
Fixed

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_ingress_test.go:
```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
2019 :stuck_out_tongue_winking_eye:

amimof replied:
Fixed

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_ingress_test.go:
```diff
@@ -0,0 +1,142 @@
+	*fake.RESTClient
+}
+
+func (c *IngressRESTClient) Post() *restclient.Request {
```
That seems overly complicated. Have a look at other create command and their testing suites.

amimof replied:
Looked at other commands as you suggested and got this sorted out

soltysh commented on staging/src/k8s.io/kubectl/pkg/generate/versioned/ingress.go:
```diff
@@ -0,0 +1,128 @@
+import (
+	"fmt"
+
+	//"strings"
```
Remove what's not used.

amimof replied:
Fixed

soltysh commented:
Also, pls squash all into a single commit.

amimof commented:
/remove-lifecycle stale

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_ingress.go:
```diff
@@ -0,0 +1,215 @@
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/cli-runtime/pkg/resource"
+	v1beta1client "k8s.io/client-go/kubernetes/typed/extensions/v1beta1"
```
Use networking group here, extensions is already deprecated.

amimof replied:
Fixed

soltysh commented on staging/src/k8s.io/kubectl/pkg/generate/versioned/generator.go:
```diff
@@ -60,6 +60,7 @@ const (
 	PriorityClassV1Alpha1GeneratorName      = "priorityclass/v1alpha1"
 	PriorityClassV1Beta1GeneratorName       = "priorityclass/v1beta1"
 	PriorityClassV1GeneratorName            = "priorityclass/v1"
+	IngressV1GeneratorName									= "extensions/v1beta1"
```
Not needed, drop it.

amimof replied:
Fixed

soltysh commented on staging/src/k8s.io/kubectl/pkg/generate/versioned/ingress.go:
```diff
@@ -0,0 +1,124 @@
+/*
```
Not needed, drop it.

amimof replied:
I don't understand 😕Dropping the first line would break the comment?

soltysh replied:
This file is not needed, since you have dedicated method in create ingress.

soltysh commented on staging/src/k8s.io/kubectl/pkg/generate/versioned/ingress_test.go:
```diff
@@ -0,0 +1,128 @@
+/*
```
Not needed, drop it.

soltysh reviewed (approved):
/lgtm
/approve
/ok-to-test

amimof commented:
/retest

amimof commented:
/retest

amimof commented:
/test pull-kubernetes-e2e-kind

amimof commented:
@soltysh @seans3 Alla tests have passed, just need lgtm label. Can you please check? Thanks

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_ingress.go:
```diff
@@ -0,0 +1,215 @@
+import (
+	"context"
+	"fmt"
+	"github.com/spf13/cobra"
```
Please split imports into following sections:
1. builtin go
2. 3-rd party
3. k8s
See other files, for example.

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_ingress.go:
```diff
@@ -0,0 +1,215 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
2020

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_ingress.go:
```diff
@@ -0,0 +1,215 @@
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/cli-runtime/pkg/resource"
+	v1beta1client "k8s.io/client-go/kubernetes/typed/networking/v1beta1"
```
`networkingv1beta1client` is preferred alias

soltysh replied:
Also there's `networking/v1` now, why not using that?

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_ingress.go:
```diff
@@ -0,0 +1,215 @@
+	}
+	if len(o.ServiceName) == 0 {
+		return fmt.Errorf("--service-name must be specified")
+	}
```
You don't need to repeat since you have those marked required in flags.

rikatz commented:
Hello @amimof 

We've returned the discussion of kubectl create ingress due to issue #93267 and I've been working in this in the PR https://github.com/kubernetes/kubernetes/pull/94327

First of all, I'm sorry I didn't saw your PR before starting mine :/

Anyway we got some discussions in sig-network mailing list of what should be the command syntax, and I've developed with that in mind.

So I want to know if you want to move forward with your PR, if we should just close it or otherwise if you want to sync both works (mine and yours) you can ping me in Slack :)

Thank you!

soltysh commented on go.mod:
```diff
@@ -21,8 +21,8 @@ require (
-	github.com/aws/aws-sdk-go v1.28.2
-	github.com/blang/semver v3.5.0+incompatible
+	github.com/aws/aws-sdk-go v1.31.12
+	github.com/blang/semver v3.5.1+incompatible
```
You should not modify any of these, not necessary in your PR.

soltysh reviewed (changes requested):
Drop go.mod related changes and this can merge

amimof commented:
After discussion with @rikatz we decided to merge work done in both PR's, this and https://github.com/kubernetes/kubernetes/pull/94327

amimof committed:
Add `create ingress` command to `cmd/kubectl`


Add `create ingress` unit tests


Move src code to staging dir


Update create command to reflect new API


Replaced deprecated `extensions` api with `networking`


Fix `missing strict dependencies`


Update BUILD 


Update BUILD


Fix commit conflict with upstream


Update after review

* Removed obsolete files
* Moved v1beta to v1 api

Fixed gofmt


Fixed deps imports


Merge with PR #94327


Revert changes


Revert go.mod


Revert BUILD


No need to update generated BUILD


Add required deps to BUILD


Update BUILD

soltysh reviewed (approved):
/lgtm

soltysh commented:
/priority backlog
